util/tempfile_test.go

```diff
@@ -48,6 +50,53 @@ func (t *helperTracker) Cleanup(f func()) {
 	t.t.Cleanup(f)
 }
 
+func trackFailure(t util.T) *failureTracker {
+	return &failureTracker{t: t}
+}
+
+type failureTracker struct {
+	failed bool
+	log    bytes.Buffer
+	t      util.T
+}
+
+func (t *failureTracker) TempDir() string {
+	t.t.Helper()
+	return t.t.TempDir()
+}
+
+func (t *failureTracker) Helper() {
+	t.t.Helper()
+}
+
+func (t *failureTracker) Errorf(s string, args ...any) {
+	t.t.Helper()
+	t.failed = true
+	fmt.Fprintf(&t.log, s+"\n", args...)
+}
+
+func (t *failureTracker) Fatalf(s string, args ...any) {
+	t.t.Helper()
+	t.failed = true
+	fmt.Fprintf(&t.log, s+"\n", args...)
+}
+
+func (t *failureTracker) Cleanup(f func()) {
+	t.t.Helper()
+	t.t.Cleanup(f)
+}
+
+func (t *failureTracker) AssertFailedWith(msg string) {
+	t.t.Helper()
+	if !t.failed {
+		t.t.Fatalf("expected test to fail with message %q", msg)
+	}
+	strlog := t.log.String()
+	if !strings.Contains(strlog, msg) {
+		t.t.Fatalf("expected test to fail with message %q\ngot message %q", msg, strlog)
+	}
+}
```

Comment:
This would be easier to maintain  and read I think

```suggestion
func (t *failureTracker) TempDir() string {
	t.Helper()
	return t.t.TempDir()
}

func (t *failureTracker) Helper() {
	t.t.Helper()
}

func (t *failureTracker) Errorf(s string, args ...any) {
	t.Helper()
	t.failed = true
	fmt.Fprintf(&t.log, s+"\n", args...)
}

func (t *failureTracker) Fatalf(s string, args ...any) {
	t.Helper()
	t.failed = true
	fmt.Fprintf(&t.log, s+"\n", args...)
}

func (t *failureTracker) Cleanup(f func()) {
	t.Helper()
	t.t.Cleanup(f)
}

func (t *failureTracker) AssertFailedWith(msg string) {
	t.Helper()
	if !t.failed {
		t.t.Fatalf("expected test to fail with message %q", msg)
	}
	strlog := t.log.String()
	if !strings.Contains(strlog, msg) {
		t.t.Fatalf("expected test to fail with message %q\ngot message %q", msg, strlog)
	}
}
```

Reply:
Good catch, thanks!